memory.c

```diff
@@ -68,30 +90,136 @@ void memory_destroy(memory mem) {
 }
 
 int memory_read_byte(memory mem, uint32_t address, uint8_t *value) {
-    if (!mem) {
-        return -NULL_REFERENCE;
+    _check_mem_integrity(mem);
+
+    if (address < 0 || address >= mem->size) { 
+        error("address out of range\n"); 
+        return INDEX_OUT_OF_RANGE; 
     }
 
-    return -1;
+    *value = mem->data[address];
+    return SUCCESS;
 
 }
 
 int memory_read_half(memory mem, uint32_t address, uint16_t *value, uint8_t be) {
-    return -1;
+    _check_mem_integrity(mem);
+
+    if (address < 0 || address >= mem->size-1) { 
+        error("half word address out of range\n"); 
+        return INDEX_OUT_OF_RANGE; 
+    } 
+
+    uint16_t aux = 0;
+    be = (be > 0) ? MEM_BIG_ENDIAN : MEM_LITTLE_ENDIAN; // cast be to bool
+    switch ((uint8_t)(be == is_big_endian()))   // check if be is sytem convention
+    {
+    case 0: // false inverse value
+        aux |= ((uint16_t)(mem->data[address++])) << 8;
+        aux |= (uint16_t)(mem->data[address]);
```

Comment:
Wrong inverse of endian

Reply:
You didn't change anything with your inverse.
You must push what we see first to the right not the left.
byte1-byte2 --> byte2-byte1 (we see byte1 first at address then byte2 at address + 1)
So it is like:
`aux |= (uint16_t)(mem->data[address]);
aux |= (unint16_t)(mem->data[address+1]) << 8;`